static/js/src/advantage/manage-subscription.js

```diff
@@ -1,93 +1,259 @@
+        retryGetPurchase(purchase_id, attempts_counter);
+
+        return;
+      }
+
+      if (invoice.pi_status === "requires_payment_method") {
+        handleError(elements, "resizing_machines_fail");
+
+        return;
+      }
+
+      if (requiresAuthentication(invoice)) {
+        authenticate_3ds(invoice, elements);
       }
-    })
-    .catch((error) => {
-      console.error(error);
     });
-}
+  }, 2000);
+};
 
-function cancelSubscription(id, VPSize) {
-  const cancelSubscriptionButton = document.querySelector(
-    `#cancel-subscription--${id}[data-viewport="${VPSize}"]`
+const requiresAuthentication = (invoice) => {
+  if (invoice.pi_decline_code === "authentication_required") {
+    return true;
+  }
+
+  if (invoice.pi_status === "requires_action" && invoice.pi_secret) {
+    return true;
+  }
+
+  return false;
+};
+
+const handleUpdateClick = (id, VPSize) => {
+  const resizeField = document.querySelector(
+    `#resize-input--${id}[data-viewport="${VPSize}"]`
+  );
+  const updateButton = document.querySelector(
+    `#save-changes--${id}[data-viewport="${VPSize}"]`
   );
+
   const {
     accountId,
     productListingId,
     previousPurchaseId,
-  } = cancelSubscriptionButton.dataset;
-  const confirmCancelButton = document.querySelector(`#confirmCancelButton`);
+    billingPeriod,
+  } = updateButton.dataset;
+
+  const successNotificationElements = document.querySelectorAll(
+    `.success-${id}`
+  );
+  const cautionNotificationElements = document.querySelectorAll(
+    `.caution-${id}`
+  );
+
+  const elements = {
+    button: {
+      html: updateButton,
+      text: "Save changes",
+    },
+    success: successNotificationElements,
+    caution: cautionNotificationElements,
+  };
 
   dataLayer.push({
     event: "GAEvent",
     eventCategory: "Advantage",
-    eventAction: "cancel-subscription",
-    eventLabel: "Cancel subscription",
+    eventAction: "update-subscription",
+    eventLabel: "Save changes",
     eventValue: undefined,
   });
 
-  handleAPICall(
-    cancelContract,
-    [accountId, previousPurchaseId, productListingId],
-    confirmCancelButton
-  );
-}
+  updateButton.classList.add("is-processing");
+  updateButton.innerHTML =
+    '<i class="p-icon--spinner u-animation--spin is-light"></i>';
+  updateButton.disabled = true;
 
-function handleUpdateClick(id, VPSize) {
-  const resizeField = document.querySelector(
-    `#resize-input--${id}[data-viewport="${VPSize}"]`
-  );
-  const updateButton = document.querySelector(
-    `#save-changes--${id}[data-viewport="${VPSize}"]`
+  resizeContract(
+    accountId,
+    previousPurchaseId,
+    productListingId,
+    resizeField.value,
+    billingPeriod
+  ).then((data) => {
+    if (data.errors) {
+      if (data.errors.includes("can only make one purchase")) {
+        handleError(elements, "pending_purchase");
+
+        setTimeout(() => {
+          location.reload();
+        }, 2000);
+
+        return;
+      }
+
```

Comment:
```suggestion

       } else {

```

use else instead of return statements

Reply:
One day we will sit down and how a chat how `else` is the root of everything evil in the IT world, causing pointless indentation and nesting. But in this case it does make the code look better. I guess JS looks so bad even an `else` can makes it prettier. :D

I am kidding, but I wish I could just to `return handleError()` so my quick returns looks nicer. I left my quick return only inside the `getPurchase` request where I think makes the code better, but if you think I should change it there, let me know.